tests/testfont.c

```diff
@@ -823,6 +826,314 @@ static void test_getLogfontW ()
 	GdipDisposeImage ((GpImage *) image);
 }
 
+static const WCHAR Tahoma[] = {'T', 'a', 'h', 'o', 'm', 'a', 0};
+static const WCHAR CodeNewRoman[] = {'C', 'o', 'd', 'e', ' ', 'N', 'e', 'w', ' ', 'R', 'o', 'm', 'a', 'n', 0};
+static const WCHAR NoSuchFont[] = {'A', 'B', 'C', 0};
+
+static void verifyFontFamily (GpFontFamily *family, const char *expectedName, UINT16 expectedCellAscent, UINT16 expectedCellDescent, UINT16 expectedLineSpacing, const char *alternateExpectedName, UINT16 alternateCellAscent, UINT16 alternateCellDescent, UINT16 alternateLineSpacing)
+{
+	GpStatus status;
+	WCHAR name[LF_FACESIZE];
+	UINT16 emHeight;
+	UINT16 cellAscent;
+	UINT16 cellDescent;
+	UINT16 lineSpacing;
+
+	status = GdipGetFamilyName (family, name, 1);
+	assert (status == Ok);
+	assert (stringsEqual (name, expectedName) || stringsEqual (name, alternateExpectedName));
+
+	status = GdipGetEmHeight (family, FontStyleRegular, &emHeight);
+	assert (status == Ok);
+	assert (emHeight == 2048);
+
+	status = GdipGetCellAscent (family, FontStyleRegular, &cellAscent);
+	assert (status == Ok);
+	assert (cellAscent == expectedCellAscent || cellAscent == alternateCellAscent);
+
+	status = GdipGetCellDescent (family, FontStyleRegular, &cellDescent);
+	assert (status == Ok);
+	assert (cellDescent == expectedCellDescent || cellDescent == alternateCellDescent);
+
+	status = GdipGetLineSpacing (family, FontStyleRegular, &lineSpacing);
+	assert (status == Ok);
+	assert (lineSpacing == expectedLineSpacing || lineSpacing == alternateLineSpacing);
+
+	GdipDeleteFontFamily (family);
+}
+
+static void test_createFontFamilyFromName ()
+{
+	GpStatus status;
+	GpFontFamily *family;
+	GpFontCollection *collection;
+	WCHAR *fontName;
+
+	GdipNewPrivateFontCollection (&collection);
+	fontName = g_utf8_to_utf16 ("test.ttf", -1, NULL, NULL, NULL);
+	GdipPrivateAddFontFile (collection, fontName);
+
+	status = GdipCreateFontFamilyFromName (Tahoma, NULL, &family);
+	assert (status == Ok);
+	verifyFontFamily (family, "Tahoma", 2049, 423, 2472, "DejaVu Sans", 1901, 483, 2384);
+
+	status = GdipCreateFontFamilyFromName (CodeNewRoman, collection, &family);
+	assert (status == Ok);
+	verifyFontFamily (family, "Code New Roman", 2059, 430, 2489, "Code New Roman", 1901, 483, 2384);
```

Comment:
@hughbe Where do these values come from? I have been doing some text with macOS+Pango (--with-pango option) and on Windows. In both cases the reported values for ascent and descent were 1884 and 514, which are specified in the font in the TT_OS2 table in usWinAscent and usWinDescent fields. Maybe I am missing something since the test obviously passes on some platforms with the cairo backend.

Reply:
Oh, it actually fails only in the macOS + Pango + CoreText backend. Otherwise the metrics are calculated from FreeType in the same way as the default Cairo backend.

Reply:
The font tests do need a bit of work - most of the currently fail on Windows against libgdiplus.

As per where the values came from, I ran the tests against GDI+ and inputted the values returned from GDI+

Reply:
That's the weird thing, because there are two sets of values. I expected one of the to match current libgdiplus cairo implementation and the other to match whatever GdiPlus.dll returns on Windows. That just doesn't happen on my machine though.

Reply:
Turns out that the first set of values is actually for Tahoma. Both the default cairo backend and pango w/ FC backend incorrectly maps to Tahoma instead of Code New Roman.

Reply:
That's a pretty bad, right? CodeNewRoman should be fetched from the font collection... shame I didn't put more care into testing this

Reply:
Yep, it's a bug. On Cairo backend it stems from the cairo_ft_font_face_create_for_pattern call, which ignore the private font collections.